splash_ingest/ingestors.py

```diff
@@ -253,6 +262,44 @@ def calc_num_events(mapping_fields: List[StreamMappingField], file):
     except KeyError as e:
         raise FieldNotInResourceError("timestamp check", name)
 
+
+def _log_and_auto_contrast(data):
+    log_image = data - np.min(data) + 1.001
+    log_image = np.log(log_image)
+    log_image = 205*log_image/(np.max(log_image))
+    auto_contrast_image = Image.fromarray(log_image.astype('uint8'))
+    auto_contrast_image = ImageOps.autocontrast(
+                            auto_contrast_image, cutoff=0.1)
+    return np.asarray(auto_contrast_image)
+
+
+class _ImageWithCustomScaling(auto_plot_builders._ShimmedImage):
+
+    def _transform(self, run, field):
+        # I am not 100% sure this is the best language feature to use for injecting
+        # a custom transform but otherwise reusing the rest of the logic in Images.
+        # That's why this method is currently private. This will either become
+        # public or we'll choose a different mechanism. - Dan Allan, Dec 2020
+
+        # The base class gives us a 2D image by slicing the middle of any
+        # higher dimensions.
```

Comment:
Elaborating on this comment:

In the base class `__init__` we are effectively doing "select" (which columns) and "combine" (make an array of the right shape to plot). If you want to leave that logic as it is but add additional reduction / cleaning / whatever, you have to subclass override a method that I'm not even sure should exist. That feels extreme; there ought to be a gentler extension mechanism for something this simple.

The use case here might be described as inserting a step in the data processing pipeline on the way to plotting.

Reply:
After the event_model fix is released, will this need to be removed? Is the summary of that fix that is we never need to close h5 files explicitly? That when the file is GC'd, it gets closed auto-magically? Is the close method in h5py really just intended for ending write access?

Reply:
Will continue this thread near the relevant lines.

Reply:
Reflections on this:
1. I should add a first-class concept of (auto-)scaling to bluesky-widgets' `Image` plot builder. It was useful for Splash to point that out.
2. I think the right way to formulate it is a manipulating on the scaling (log vs linear) and range (min, max) of the color bar, not by changing the array itself. We should hand the actual data to matplotlib (or whatever plotting framework) so that when there is a colorbar, its scale represents physically-meaning values.
3. I still feel unsure about how best to make it easy to extend whatever built-in heuristics and/or reduction bluesky-widgets built-in classes do. I do not think subclassing is the way.

Reply:
Yeah, it seems to me that in some future version, it might be nice to think of handy reusable transform helpers that could be plugged in in spots like this. Heck, I don't even remember where I got the log_and_auto_contrast code originally. :)